Subject: [PATCH] fix -Wall -Wextra -Werror warnings

Unused-parameter (void) casts in xen/zlib/zstd stubs, sign-compare
casts, warn_unused_result checks (ftruncate/fgets/read),
fallthrough comments in getopt switches, drop dead counter i.
Purgatory not touched (freestanding, -fno-builtin).
diff --git a/kexec/arch/i386/crashdump-x86.c b/kexec/arch/i386/crashdump-x86.c
index 0716762..02e7175 100644
--- a/kexec/arch/i386/crashdump-x86.c
+++ b/kexec/arch/i386/crashdump-x86.c
@@ -394,6 +394,9 @@ err:
 static int get_crash_memory_ranges_xen(struct memory_range **range,
 					int *ranges, unsigned long lowmem_limit)
 {
+	(void)range;
+	(void)ranges;
+	(void)lowmem_limit;
 	return 0;
 }
 #endif /* HAVE_LIBXENCTRL */
diff --git a/kexec/arch/i386/kexec-beoboot-x86.c b/kexec/arch/i386/kexec-beoboot-x86.c
index d949ab8..f60be98 100644
--- a/kexec/arch/i386/kexec-beoboot-x86.c
+++ b/kexec/arch/i386/kexec-beoboot-x86.c
@@ -103,6 +103,7 @@ int beoboot_load(int argc, char **argv, const char *buf, off_t UNUSED(len),
 			if (opt < OPT_ARCH_MAX) {
 				break;
 			}
+		/* fall through */
 		case OPT_REAL_MODE:
 			real_mode_entry = 1;
 			break;
diff --git a/kexec/arch/i386/kexec-bzImage.c b/kexec/arch/i386/kexec-bzImage.c
index 81ca2df..ff52854 100644
--- a/kexec/arch/i386/kexec-bzImage.c
+++ b/kexec/arch/i386/kexec-bzImage.c
@@ -219,9 +219,9 @@ int do_bzImage_load(struct kexec_info *info,
 			kern16_size_needed = setup_header.heap_end_ptr;
 		else
 			kern16_size_needed = kern16_size + 8192; /* bss */
-		if (kern16_size_needed < kern16_size)
+		if (kern16_size_needed < (unsigned long)kern16_size)
 			kern16_size_needed = kern16_size;
-		if (kern16_size_needed > 0xfffc)
+		if (kern16_size_needed > 0xfffcUL)
 			die("kern16_size_needed is more then 64k\n");
 		heap_size = 0xfffc - kern16_size_needed; /* less 64k */
 		heap_size = _ALIGN_DOWN(heap_size, 0x200);
@@ -428,6 +428,7 @@ int bzImage_load(int argc, char **argv, const char *buf, off_t len,
 			if (opt < OPT_ARCH_MAX) {
 				break;
 			}
+		/* fall through */
 		case OPT_APPEND:
 			append = optarg;
 			break;
diff --git a/kexec/arch/i386/kexec-elf-x86.c b/kexec/arch/i386/kexec-elf-x86.c
index 8eba242..2282633 100644
--- a/kexec/arch/i386/kexec-elf-x86.c
+++ b/kexec/arch/i386/kexec-elf-x86.c
@@ -161,6 +161,7 @@ int elf_x86_load(int argc, char **argv, const char *buf, off_t len,
 			if (opt < OPT_ARCH_MAX) {
 				break;
 			}
+		/* fall through */
 		case OPT_APPEND:
 			append = optarg;
 			break;
diff --git a/kexec/arch/i386/kexec-mb2-x86.c b/kexec/arch/i386/kexec-mb2-x86.c
index 0d2e93b..c96a1bc 100644
--- a/kexec/arch/i386/kexec-mb2-x86.c
+++ b/kexec/arch/i386/kexec-mb2-x86.c
@@ -84,11 +84,11 @@ int multiboot2_x86_probe(const char *buf, off_t buf_len)
 
 	/* Now look for a multiboot header. */
 	len = MULTIBOOT_SEARCH;
-	if (len > buf_len)
+	if ((off_t)len > buf_len)
 		len = buf_len;
 
 	memcpy(headerbuf, buf, len);
-	if (len < sizeof(struct multiboot_header)) {
+	if ((size_t)len < sizeof(struct multiboot_header)) {
 		/* Short file */
 		return -1;
 	}
@@ -377,6 +377,7 @@ out:
 static uint64_t multiboot2_mbi_add_module(void *mbi_buf, uint64_t mbi_ptr, uint32_t mod_start,
 					  uint32_t mod_end, char *mod_clp)
 {
+	(void)mbi_buf;
 	struct multiboot_tag_module *tag = (struct multiboot_tag_module *) (uintptr_t) mbi_ptr;
 
 	tag->type = MULTIBOOT_TAG_TYPE_MODULE;
@@ -392,6 +393,7 @@ static uint64_t multiboot2_mbi_add_module(void *mbi_buf, uint64_t mbi_ptr, uint3
 
 static uint64_t multiboot2_mbi_end(void *mbi_buf, uint64_t mbi_ptr)
 {
+	(void)mbi_buf;
 	struct multiboot_tag *tag = (struct multiboot_tag *) (uintptr_t) mbi_ptr;
 
 	tag->type = MULTIBOOT_TAG_TYPE_END;
@@ -459,6 +461,7 @@ int multiboot2_x86_load(int argc, char **argv, const char *buf, off_t len,
 			if (opt < OPT_ARCH_MAX) {
 				break;
 			}
+		/* fall through */
 		case OPT_CL:
 			append = optarg;
 			break;
@@ -559,8 +562,6 @@ int multiboot2_x86_load(int argc, char **argv, const char *buf, off_t len,
 	if (modules) {
 		char *mod_filename, *mod_command_line, *mod_clp, *buf;
 		off_t mod_size;
-		int i = 0;
-
 		mod_clp_base = xmalloc(mod_command_line_space);
 
 		/* Go back and parse the module command lines */
@@ -592,7 +593,6 @@ int multiboot2_x86_load(int argc, char **argv, const char *buf, off_t len,
 			mbi_ptr = multiboot2_mbi_add_module(mbi_buf, mbi_ptr, addr, addr + mod_size, mod_clp);
 
 			mod_clp += strlen(mod_clp) + 1;
-			i++;
 		}
 
 		free(mod_clp_base);
diff --git a/kexec/arch/i386/kexec-multiboot-x86.c b/kexec/arch/i386/kexec-multiboot-x86.c
index 33c885a..434cbe3 100644
--- a/kexec/arch/i386/kexec-multiboot-x86.c
+++ b/kexec/arch/i386/kexec-multiboot-x86.c
@@ -269,6 +269,7 @@ int multiboot_x86_load(int argc, char **argv, const char *buf, off_t len,
 			if (opt < OPT_ARCH_MAX) {
 				break;
 			}
+		/* fall through */
 		case OPT_CL:
 			append = optarg;
 			break;
diff --git a/kexec/arch/i386/kexec-x86-common.c b/kexec/arch/i386/kexec-x86-common.c
index 4a67ec8..d027195 100644
--- a/kexec/arch/i386/kexec-x86-common.c
+++ b/kexec/arch/i386/kexec-x86-common.c
@@ -227,6 +227,8 @@ err:
 #else
 static int get_memory_ranges_xen(struct memory_range **range, int *ranges)
 {
+	(void)range;
+	(void)ranges;
 	return 0;
 }
 #endif /* HAVE_LIBXENCTRL */
@@ -317,7 +319,8 @@ int efi_map_added( void ) {
 	char buf[512];
 	FILE *fp = fopen( "/proc/cmdline", "r" );
 	if( fp ) {
-		fgets( buf, 512, fp );
+		if( fgets( buf, sizeof(buf), fp ) == NULL )
+			buf[0] = '\0';
 		fclose( fp );
 		return strstr( buf, "add_efi_memmap" ) != NULL;
 	} else {
diff --git a/kexec/arch/i386/x86-linux-setup.c b/kexec/arch/i386/x86-linux-setup.c
index 94d9dd2..0b856e1 100644
--- a/kexec/arch/i386/x86-linux-setup.c
+++ b/kexec/arch/i386/x86-linux-setup.c
@@ -103,7 +103,7 @@ void setup_linux_bootloader_parameters_high(
 		real_mode->ext_ramdisk_image = (initrd_base >> 16) >> 16;
 
 	if (real_mode->protocol_version >= 0x020c &&
-	    (initrd_size & 0xffffffffUL) != initrd_size)
+	    ((unsigned long)initrd_size & 0xffffffffUL) != (unsigned long)initrd_size)
 		real_mode->ext_ramdisk_size = (initrd_size >> 16) >> 16;
 
 	/* The location of the command line */
@@ -511,7 +511,8 @@ int get_bootparam(void *buf, off_t offset, size_t size)
 		return 1;
 	if (lseek(data_file, offset, SEEK_SET) < 0)
 		goto close;
-	read(data_file, buf, size);
+	if (read(data_file, buf, size) < 0)
+		goto close;
 close:
 	close(data_file);
 	return 0;
diff --git a/kexec/arch/x86_64/kexec-bzImage64.c b/kexec/arch/x86_64/kexec-bzImage64.c
index 29ed8c9..046dc6f 100644
--- a/kexec/arch/x86_64/kexec-bzImage64.c
+++ b/kexec/arch/x86_64/kexec-bzImage64.c
@@ -272,6 +272,7 @@ int bzImage64_load_file(int argc, char **argv, struct kexec_info *info)
 			/* Ignore core options */
 			if (opt < OPT_ARCH_MAX)
 				break;
+		/* fall through */
 		case OPT_APPEND:
 			append = optarg;
 			break;
@@ -361,6 +362,7 @@ int bzImage64_load(int argc, char **argv, const char *buf, off_t len,
 			/* Ignore core options */
 			if (opt < OPT_ARCH_MAX)
 				break;
+		/* fall through */
 		case OPT_APPEND:
 			append = optarg;
 			break;
diff --git a/kexec/arch/x86_64/kexec-elf-x86_64.c b/kexec/arch/x86_64/kexec-elf-x86_64.c
index 7f9540a..87c6cba 100644
--- a/kexec/arch/x86_64/kexec-elf-x86_64.c
+++ b/kexec/arch/x86_64/kexec-elf-x86_64.c
@@ -107,6 +107,7 @@ int elf_x86_64_load(int argc, char **argv, const char *buf, off_t len,
 				break;
 			}
 			fprintf(stderr, "Unknown option: opt: %d\n", opt);
+			/* fall through */
 		case OPT_APPEND:
 			append = optarg;
 			break;
diff --git a/kexec/crashdump-xen.c b/kexec/crashdump-xen.c
index 3f59a0d..3c1381f 100644
--- a/kexec/crashdump-xen.c
+++ b/kexec/crashdump-xen.c
@@ -32,6 +32,7 @@ static int is_dom0;
 static jmp_buf xen_sigill_jmp;
 void xen_sigill_handler(int sig)
 {
+	(void)sig;
 	longjmp(xen_sigill_jmp, 1);
 }
 
@@ -218,6 +219,8 @@ out:
 #else
 int get_xen_vmcoreinfo(uint64_t *addr, uint64_t *len)
 {
+	(void)addr;
+	(void)len;
 	return -1;
 }
 
@@ -251,6 +254,8 @@ int xen_get_crashkernel_region(uint64_t *start, uint64_t *end)
 #else
 int xen_get_crashkernel_region(uint64_t *start, uint64_t *end)
 {
+	(void)start;
+	(void)end;
 	return -1;
 }
 #endif
diff --git a/kexec/kexec-pe-zboot.c b/kexec/kexec-pe-zboot.c
index c09f2ae..66d2231 100644
--- a/kexec/kexec-pe-zboot.c
+++ b/kexec/kexec-pe-zboot.c
@@ -112,7 +112,8 @@ int pez_prepare(const char *crude_buf, off_t buf_sz, int *kernel_fd,
 		/* If signed, the Attribute Certificate Table is always at the end of the PE file */
 		if (dir->certs.virtual_address != 0 && dir->certs.size != 0) {
 			original_file_sz = dir->certs.virtual_address + dir->certs.size;
-			ftruncate(fd, 0);
+			if (ftruncate(fd, 0))
+				dbgprintf("%s: ftruncate failed\n", __func__);
 		}
 	}
 
diff --git a/kexec/kexec-uki.c b/kexec/kexec-uki.c
index 9888d7e..75943c9 100644
--- a/kexec/kexec-uki.c
+++ b/kexec/kexec-uki.c
@@ -50,6 +50,7 @@ static int create_tmpfd(const char *template, char *buf, int buf_sz, int *tmpfd)
 
 int uki_image_probe(const char *file_buf, off_t buf_sz)
 {
+	(void)buf_sz;
 	struct pe_hdr *pe_hdr;
 	struct pe32plus_opt_hdr *opt_hdr;
 	struct section_header *sect_hdr;
diff --git a/kexec/kexec-xen.c b/kexec/kexec-xen.c
index a7c8933..abb2aed 100644
--- a/kexec/kexec-xen.c
+++ b/kexec/kexec-xen.c
@@ -271,6 +271,9 @@ int xen_kexec_exec(uint64_t kexec_flags)
 
 int xen_get_kexec_range(int range, uint64_t *start, uint64_t *end)
 {
+	(void)range;
+	(void)start;
+	(void)end;
 	return -1;
 }
 
@@ -281,16 +284,19 @@ int xen_kexec_load(struct kexec_info *UNUSED(info))
 
 int xen_kexec_unload(uint64_t kexec_flags)
 {
+	(void)kexec_flags;
 	return -1;
 }
 
 int xen_kexec_status(uint64_t kexec_flags)
 {
+	(void)kexec_flags;
 	return -1;
 }
 
 int xen_kexec_exec(uint64_t kexec_flags)
 {
+	(void)kexec_flags;
 	return -1;
 }
 
diff --git a/kexec/kexec.c b/kexec/kexec.c
index c9e4bcb..2682471 100644
--- a/kexec/kexec.c
+++ b/kexec/kexec.c
@@ -1187,7 +1187,7 @@ static ssize_t _read(int fd, void *buf, size_t count)
 			return ret;
 		}
 		offset += ret;
-	} while (ret && offset < count);
+	} while (ret && (size_t)offset < count);
 
 	return offset;
 }
diff --git a/kexec/zlib.c b/kexec/zlib.c
index 3ed6bd6..c46450f 100644
--- a/kexec/zlib.c
+++ b/kexec/zlib.c
@@ -25,6 +25,7 @@ static void _gzerror(gzFile fp, int *errnum, const char **errmsg)
 
 int is_zlib_file(const char *filename, off_t *r_size)
 {
+	(void)r_size;
 	gzFile fp;
 	int errnum;
 	int is_zlib_file = 0; /* default: It's not in gzip format */
diff --git a/kexec/zstd.c b/kexec/zstd.c
index eaa76b9..7936bc3 100644
--- a/kexec/zstd.c
+++ b/kexec/zstd.c
@@ -24,6 +24,7 @@
 __attribute__((weak))
 unsigned ZSTD_isFrame(const void* buffer, size_t size)
 {
+	(void)size;
 	uint8_t *buf = (uint8_t *)buffer;
 	/* Magic zstd frame header value */
 	if ((buf[0] == 0x28) &&
@@ -80,7 +81,7 @@ char *zstd_decompress_file(const char *filename, off_t *r_size)
 		dbgprintf("Cannot stat `%s'\n", filename);
 		return NULL;
 	}
-	if (fp_stats.st_size < sizeof(magic)) {
+	if ((size_t)fp_stats.st_size < sizeof(magic)) {
 		dbgprintf("short file\n");
 		return NULL;
 	}
